### DISCUSSION TITLE
Default device management

### DISCUSSION THREAD

**Discussion by ivoadf** [wwmm/easyeffects#1836] [General]
I'm a bit confused with the following guidance under the pipewire tab
![image](https://user-images.githubusercontent.com/9215666/190914903-43168848-9bc7-4a6a-a580-d59729b62d72.png)

My expectation would be that the easy effects sync/source would be the default system wide so that all applications would use them forcing any effects to be applied. If we don't do this how else would applications get the effects applied to their sync and source audio?

Would you mind explaining the reasoning behind this hint? I am probably misunderstanding the correct usage of this.

Thanks

**Comment by wwmm** [comment]
>My expectation would be that the easy effects sync/source would be the default system wide so that all applications would use them forcing any effects to be applied.

In its default configuration EasyEffects tries to use the default system input/output device. If you set our own devices as default there is a chance that in EasyEffects initialization phase PipeWire will report our own devices as the default devices and as EasyEffects can't play to / record from itself it won't know anymore which device it should use. This has happened to some users.

>If we don't do this how else would applications get the effects applied to their sync and source audio?

There is no need to set its virtual devices as default. For example when `Process All Output Streams` is enabled in our configuration window EasyEffects will move streams to its virtual output on its own. Except in the corner cases where the audio app developers decided to set a server flag that forbids third party tools from moving their app streams. There is nothing I can do when developers misuse the server flags.

**Reply by ivoadf** [reply]
Ok, so I was unaware that easy effects can "hijack" streams.
Just to make sure I understand let me run through an example.
In the scenario where I want all my applications to use the default device with effects applied both for sync and source I can just configure the default sync/source that I want and toggle the "use default" options in easy effects and it will "hijack" the streams between the application and the default device to apply the necessary effects.
Is this correct?

**Reply by wwmm** [reply]
>Is this correct?

Yes. If you have the following settings enabled

- Use Default input
- Use Default Output
- Process All Inputs Streams
- Process All Outputs Streams

EasyEffects will use the default input/output system device and will move the audio application streams to its virtual input/output. Except in the corner cases where the audio app developer committed the cardinal sin of disabling the stream management. Fortunately the vast majority of the developers do not do this.

In case there is a particular app you do not want to be touched by EasyEffects you just have to put it in the corresponding blocklist.

**Reply by ivoadf** [reply]
Awesome, now it makes sense.
Thanks for the quick reply!

**Reply by AidanofVT** [reply]
When I set an application's audio destination to the default in my OS sound utility, that disables that application in EasyEffects/Output/Players. Likewise, When I check/uncheck "Enable" in that menu, it actually changes whether or not EasyEffects is that application's audio destination in the OS. What's going on here?

**Reply by wwmm** [reply]
@AidanofVT is EasyEffects configured to process all outputs? If yes and the application dos not disable audio redirection the desired stream should be linked to EasyEffects virtual sink. 

The enable/ disable checkbox will change the device the audio stream is linked to. When enabled it will be linked to our virtual sink. If disabled the streams will most likely use the default system device.

**Reply by dlipovetsky** [reply]
For a year, I used EasyEffects as my default "virtual" sink/source. Every time my "physical" sink changed, I would open the EasyEffects UI and configure it to use the correct sink.

Recently, EasyEffects returned to configuring its PipeWire devices such that their volume could not be changed. As I tried to understand this problem, I finally understood, from the help documentation, that I could configure EasyEffects as described in https://github.com/wwmm/easyeffects/discussions/1836#discussioncomment-3675066.

Now I can change my "physical" sink without having to re-configure EasyEffects.

This thread is the first place I've seen a complete description of the  "process all streams, regardless of the input/output I use" use case, although I think it's a common use case.

**Reply by adueppen** [reply]
> Except in the corner cases where the audio app developers decided to set a server flag that forbids third party tools from moving their app streams.

By any chance is Mutter one of these? I'm currently using EasyEffects to, among other things, lower the level of my headphone jack output quite a bit, but I've noticed that the GNOME "alert sound" (what gets played when you backspace in an empty text box for example) and other sounds being produced by Mutter completely bypass EasyEffects and go straight to the output which is unpleasantly loud.
![image](https://github.com/user-attachments/assets/7bb1a5b0-c4e1-4caf-8d56-863b66829c61)

**Reply by wwmm** [reply]
>By any chance is Mutter one of these? I'm currently using EasyEffects to, among other things, lower the level of my headphone jack output quite a bit, but I've noticed that the GNOME "alert sound" (what gets played when you backspace in an empty text box for example) and other sounds being produced by Mutter completely bypass EasyEffects and go straight to the output which is unpleasantly loud.

No what you are seeing is #3131. In the past system notifications caused some undesirable side effects so we blocklisted them. I think that nowadays it should be fine to add an option to allow them to be processed. I just did not have time for that.

**Comment by polarathene** [comment]
Just chiming in here too as I found the recommendation in settings a little confusing too. It says "external applications", but the examples given seem to be OS audio settings specific, rather than general applications?

It probably would be helpful if there was a "More info" link/button that expanded some text or directed me to a resource that expanded on what Easy Effects is doing here?

What happens if the default device changes here, such as my wireless headphones example below. I could explicitly choose the headphones instead of the "Use Default Output" switch being enabled, but not sure what happens when that device is not present (_due to powering them off_), presumably Easy Effects falls back to another output device, or the default?

If I explicitly choose one, and that device is also a default device elsewhere, is the hijacking behind the scenes still the same? Can apps use the default device, or would they need to explicitly choose that device as well if not choosing an implicit default device via the UI switch/toggle button?

---

The Discord app lists that `Deity-VO-7U` mic as the selected input/default, and I'm guessing an app should either use the Default device if possible, or use the explicit `Deity-VO-7U` below it and in this case both should be the same as "Easy Effects Source" to a listener? Easy Effects is doing the hijacking thing?

![Screenshot_20221108_125451](https://user-images.githubusercontent.com/5098581/200440597-24907929-01b0-45e0-98b1-02a865242a4c.png)

So the Easy Effects source / sink should be ignored, and our apps should choose a Default device or whatever Easy Effects settings has  listed/chosen under the Input / Output "Name" values for Device Management? The Global bypass button when active presumably disables the device hijacking behind the scenes?

---

I am not sure if the Input tabs monitor button (_which I have to toggle on/off each time my bluetooth headphones are enabled to switch the monitor output to them_) is affected by the settings Default Output at all. Is it related?

Before the output was also the `Deity-VO-7U`, like the Default Input is shown above - I had toggled the switch button for Default Output, which changed the Output "Name" to my Bose headphones instead.

Perhaps the audio Output effects wasn't being applied to my headphones prior due to this, or my headphones had since become the default output device, and this was just an outdated UI value? I just know that when I power the headphones on after they've connected via bluetooth, I need to toggle the monitor button off then back on (in the Input effects view) otherwise I don't hear anything. So this may have just been UI not updating and made no real difference?

**Reply by wwmm** [reply]
>It says "external applications", but the examples given seem to be OS audio settings specific, rather than general applications?

Maybe the word external was not a good choice. But "OS tools" are not the only ones that can set default devices. Anyone that is familiar enough with PipeWire or Pulseaudio API can write a small software or script that can set default devices. In any case the intention behind the recommendation is that users `do not` set EasyEffects as default devices. Never. Always keep your hardware as default.

>What happens if the default device changes here, such as my wireless headphones example below. I could explicitly choose the headphones instead of the "Use Default Output" switch being enabled, but not sure what happens when that device is not present (due to powering them off), presumably Easy Effects falls back to another output device, or the default?

If you decide to use a custom input/output device and at some point in the future this device is removed EasyEffects will fallback to the default device. At least is what is happens on my computer when I disable one of my 2 soundcards.

>If I explicitly choose one, and that device is also a default device elsewhere, is the hijacking behind the scenes still the same? Can apps use the default device, or would they need to explicitly choose that device as well if not choosing an implicit default device via the UI switch/toggle button?

I am not sure I understand. The only real problem is using an audio manager tool to set EasyEffects virtual devices as `the system` default devices. What you do in apps device selection menus won't matter if you have settings like `Process All Outputs` enabled in EasyEffects. IF the audio app is not disabling stream device management EasyEffects will be able to redirect its audio.

>The Global bypass button when active presumably disables the device hijacking behind the scenes?

The global bypass button is for the effects in the pipeline and not the app stream redirection.

**Reply by wwmm** [reply]
>Perhaps the audio Output effects wasn't being applied to my headphones prior due to this, or my headphones had since become the default output device, and this was just an outdated UI value? I just know that when I power the headphones on after they've connected via bluetooth, I need to toggle the monitor button off then back on (in the Input effects view) otherwise I don't hear anything. So this may have just been UI not updating and made no real difference?

Bluetooth devices are frequently are pain but as far as I am aware they usually work. I think I will need to see the output file generated by `pw-dot`. This file will show how devices and filters are linked to each other and can be visualized with the command `xdot`. This way I will have a better idea about what is happening on your side.

**Comment by herrenP** [comment]
I might be daft or something. To me, at some point in this thread it sounds like only tools that are considered OS tools should set easyeffects as default sink/source. At another point it sounds like no tool ever should do this.

In any case. I am simply running a script that is toggled by a hotkey to quickly change between my speaker (connected to onboard audio card) and wireless headset (it is it's own audio card). The script uses pulseaudio's CLI tool to do this.

I have `Process All Outputs Streams` set to `true`.

```shell
case $(getRunningSink) in
    "$onboardAudioCardName")
        ...
        pactl set-default-sink $externalAudioCardName
        pactl set-default-sink easyeffects_sink
        notifyUser $headphonesPortName
        ;;
    "$externalAudioCardName")
        ...
        pactl set-default-sink $onboardAudioCardName
        pactl set-sink-port "$sinkPortNumber" $speakerPortName
        pactl set-default-sink easyeffects_sink
        notifyUser $speakerPortName
        ;;
    ...
esac
```

If I do not run `pactl set-default-sink easyeffects_sink` in the cases above I can hear that the effects are not applied. For example when I am listening to music for example in YouTube Music webapp. Should I keep doing the above or should I not run the command that set easyeffect as default sink and do something else?

I never set default sinks anywhere else than via this script.

The reason I ended up in this thread at all is because I am having an issue where, sometimes, the sound is sort of crackling when easyeffects is enabled. But I am quite sure this is related to when I am also running Android emulators in Android Studio at the same time. Those emulators seem to do stuff with sound and volume during their boot sequence.

**Reply by wwmm** [reply]
>If I do not run pactl set-default-sink easyeffects_sink in the cases above I can hear that the effects are not applied. 

This should happen only if `Process All Output Streams` is disabled in EasyEffects preferences window. Or if the app you are using is requesting the server to not let its streams to be moved to another device. If in doubt test with mpv or vlc. They do not do this kind of thing.

>The reason I ended up in this thread at all is because I am having an issue where, sometimes, the sound is sort of crackling when easyeffects is enabled. But I am quite sure this is related to when I am also running Android emulators in Android Studio at the same time. Those emulators seem to do stuff with sound and volume during their boot sequence.

They may be requesting a latency that is lower that your system can handle. Some games do this. Run `pw-top` and take a look at the `QUANT` and `RATE` columns.

**Reply by herrenP** [reply]
Alright, not setting "Easy effects sink" as the output device is working fine. My problem was probably that I had not enabled the checkboxes for the correct applications in the "Output" tab the easyeffect's configuration window.

I changed my script to set the real sinks as the default output device. But, each time I toggle between my sinks the checkboxes I mentioned above are disabled again. Reading #1757 I think I understand why. Is there a way through CLI means to enable the applications after I have toggled the default output? I will find it quite tedious to go through all listings for Chrome and find the correct one that is YouTube Music.

**Reply by wwmm** [reply]
>Is there a way through CLI means to enable the applications after I have toggled the default output? 

Enabling a stream is just "moving it to EE virtual devices". And this operation can be done through command line tools like `pactl` or `pw-metadata`. I haven't done it this way in a while but If I remember well you will have to first find the id number of the stream and also of the output device. This can also be done through pactl. Or `pw-cli list-objects Node` if you want a native PipeWire utility.

You may have to do a script that is more complex than you would want but it should work.

**Reply by wwmm** [reply]
Do you have a reason to keep `Process All Output Streams` disabled? In some cases it is easier to enable it and use the blocklist to exclude what you do not want to be redirected by EE.

**Reply by herrenP** [reply]
"Process All Output Streams" is enabled. The sound is different when I have toggled the sink output device via my script. Only when I enable the checkboxes does the sound change to what I expect.

**Reply by wwmm** [reply]
>"Process All Output Streams" is enabled.

In this case the stream `enable checkbox` should be active without the need to do it manually. Unless some kind of problem is happening to WirePlumber. We rely on the media session manager to do this operation.

**Reply by herrenP** [reply]
Ugh, this sounds like a hard thing to troubleshoot. Any idea where to look? System log for WirePlumber maybe?

Thanks for your patience!

**Reply by wwmm** [reply]
I do not remember at this moment how to put WirePlumber in debug mode. But the first thing would be to run EasyEffects in debug mode so we can have more information. But it is probably better to do this in a new issue instead of this discussion about default device management. 

In order to run EE in debug mode kill it `easyeffects -q` and restart it as `G_MESSAGES_DEBUG=easyeffects easyeffects`. Then start some audio players. under normalme circunstances the log should show they being moved to our virtual device when `Process All Output Streams` is enabled.

**Comment by teohhanhui** [comment]
Sorry if this is not the right place to ask this, but considering the other issues all link here for the advice on not setting EasyEffects sink / source as the default...

I think there should be guidance on how we can avoid the system automatically setting the EasyEffects sink / source as the default. This can happen for example when we disconnect a device. Is there a way for EasyEffects to mark as "never use as default" / deprioritize its own nodes?

**Reply by wwmm** [reply]
>I think there should be guidance on how we can avoid the system automatically setting the EasyEffects sink / source as the default.

I would love to be able to do this. But as far as I know neither PipeWire or Pulseaudio allows something like this to be done. There isn't even a flag that would say "please put this device in the lowest possible priority".

**Reply by teohhanhui** [reply]
When using WirePlumber, I think `priority.session` is supposed to help?

* https://pipewire.pages.freedesktop.org/wireplumber/configuration/alsa.html#node-properties
* https://blog.zenlinux.com/2022/08/how-to-configure-audio-device-priorities-in-pipewire-wireplumber/
* https://gitlab.freedesktop.org/pipewire/wireplumber/-/issues/352
* https://gitlab.freedesktop.org/pipewire/pipewire/-/issues/2363

**Reply by wwmm** [reply]
>When using WirePlumber, I think priority.session is supposed to help?

I do not know. Looking at the `pw-cli info object_id_number` output PipeWire does not seem to set `priority.session` for null sinks (virtual devices). This field isn't in the null-sink info output. If I force it to be set when EasyEffects loads its virtual devices then the field is there. But it is not clear if it will actually be used or just ignored. And if setting it to zero is the best approach or not.

**Reply by wwmm** [reply]
The latest release `7.1.3` sets `priority.session = 0`. Let's see if it helps WirePlumber to not set our devices automatically as default...

**Comment by violetmage** [comment]
For anyone who is new to inter-application audio or is feeling lost, I can confidently recommend [helvum](https://flathub.org/apps/org.pipewire.Helvum) as an easy visual way to interact with pipewire routing.

Edit: There is also available a more advanced app, [coppwr](https://flathub.org/apps/io.github.dimtpap.coppwr)

Anyway, here is my summary of the most common scenarios (please correct me if I'm wrong @wwmm ):

First, under no circumstances ever should easyeffects be set as the default device in your system's settings / audio mixer / whatever.

If you want easyeffects to "work like the default", set the device you want the final output to be as the default, then set easyeffects to process all outputs. Set your applications to output to the "system default" device, or whatever equivalent they offer. (Firefox, for example, will normally default to using the "system default". When easyeffects sees firefox try to set up it's output to the system default, it reroutes firefox to easyeffects virtual sink when process all outputs is enabled.)

If you want easyeffects to ignore certain applications, _do the same as above,_ and simply add the application(s) whose audio you don't want processed to the list of excluded apps.

If you want only certain applications to be processed, or want to control all the routing yourself (for example, using helvum), turn off  process all inputs/outputs in easyeffects, and _explicitly_ set the applications that you want to be processed to use easyeffects virtual sink. Again, this will make more sense  if you look at `pw-dot | xdot` or helvum.

**Comment by LinuxOnTheDesktop** [comment]
@violetmage 

I am afraid that even after reading your summary I am unsure about how to achieve the following - surely common - aim. I'd like Easy Effects to process sound that comes directly from my laptop; but, when I connect the laptop to a HiFi, or external speakers, or what have you, I'd like Easy Effects to leave the sound alone.

**Reply by wwmm** [reply]
>I am afraid that even after reading your summary I am unsure about how to achieve the following - surely common - aim. I'd like Easy Effects to process sound that comes directly from my laptop; but, when I connect the laptop to a HiFi, or external speakers, or what have you, I'd like Easy Effects to leave the sound alone.

It is not clear the workflow you are trying to do. What you can do on EasyEffects is deciding which streams you want to be redirected to its virtual devices. "Leaving the sound alone" will require that either the stream isn't moved to EasyEffects or if it is that a preset configured for `passthrough` mode is autoloaded depending on which hardware device is being used. By passthrough preset I mean a preset that has no filter added to the pipeline.

**Reply by teohhanhui** [reply]
The very minimum required for a "passthrough" mode to work is for the preset autoloading to accept wildcard devices, so that we could have no filters applied for unknown devices.

**Reply by wwmm** [reply]
Wildcard isn't required. It would just a convenience for people that have many devices. For an user that has just the main soundcard and one or two additional devices that are frequently used it is just a matter of creating autoloading profile for these additional devices. This approach only becomes annoying for people that really use many different devices frequently.

**Comment by LinuxOnTheDesktop** [comment]
@wwmm

As you can tell, I am new to this software. Still, the software is supposed to be 'easy'. :)

The leaving alone of which I spoke consists in matters proceeding, for the device (or 'stream') in question, as if Easy Effects had never existed. Thus, as you put it, 'the stream isn't moved to EasyEffects'.

**Reply by wwmm** [reply]
>The leaving alone of which I spoke consists in matters proceeding, for the device (or 'stream') in question, as if Easy Effects had never existed. Thus, as you put it, 'the stream isn't moved to EasyEffects'.

The solutions currently available will depend on the level of complexity you have in your setup and on your workflow. One possibility would be to disable `Process All Outputs` in EasyEffects window and manually enable effects for the desirable stream through the checkbutton EasyEffects show for each stream. The obvious downside is that manual intervention is required.

Assuming you do not use several different devices regularly but only 2 or 3 the easiest thing to do right now is creating profiles for preset autoloading depending on the device being used by easyeffects. Let's assume that besides the onboard card you have an external usb card and a bluetooth headphone. And that only the soundcard is supposed to have effects. While using the onboard soundcard create an autoloading profile that loads a preset with the desired effects. Now create a preset that has no filters in it. While EasyEffects is using the external card or the bluetooth headset create an autoloading profile for them that loads the preset that has no filter.

The reason why the profile can only be created while EasyEffects is using the target device is because this is the only way to know the hardware profile that is actually being used by the device.

>Still, the software is supposed to be 'easy'. :)

Ideally yes. But sometimes time is short and we have to decide where to put it. And often I favor bug fixing or new features addition. As a result the app ended a little on the "advanced user" side. But with some patience it isn't too hard for all users to get things done.

**Comment by LinuxOnTheDesktop** [comment]
> Easy Effects*
> 
> *: Title implies no guarantee that use of the application will in fact be easy.

:) I will ask for my money back!

Still, thanks for the instructions.

**Reply by wwmm** [reply]
> > Easy Effects*
> > *: Title implies no guarantee that use of the application will in fact be easy.
> 
> :) I will ask for my money back!
> 
> Still, thanks for the instructions.

Compared to setting all this up directly in the Jack server or editing PipeWire's configuration files it is still easy :smile:  hahaha

**Reply by m1st0** [reply]
I had some issues with KDE's default sync and how to get every application to behave.  So I made a WirePlumber rule routing all default audio output to EasyEffects without making it the default sync.  See my work at [WirePlumber-EasyEffects](https://github.com/m1st0/wireplumber-easyeffects) . I tested opening EasyEffects through systemd without a window, opening EasyEffects in windowed status, and things appear to work normally using FIrefox, Chrome, Vivaldi, and Zen playing a video from YouTube and as well Audacity playing an audio file.  Did you want me to add something as a pull request too? Maybe this would help others the way I have it? Let me know.

**Reply by wwmm** [reply]
>I had some issues with KDE's default sync and how to get every application to behave. So I made a WirePlumber rule routing all default audio output to EasyEffects without making it the default sync. See my work at [WirePlumber-EasyEffects](https://github.com/m1st0/wireplumber-easyeffects) . I tested opening EasyEffects through systemd without a window, opening EasyEffects in windowed status, and things appear to work normally using FIrefox, Chrome, Vivaldi, and Zen playing a video from YouTube and as well Audacity playing an audio file. Did you want me to add something as a pull request too? Maybe this would help others the way I have it? Let me know.

This feels more like a last resort when things just do not work for reasons that are hard to figure out. What fortunately does not seem to be the case for most users. For example I'm also a KDE user and things were always fine for me. And one potential issue of a WirePlumber rule that forces streams to be moved to EasyEffects sink is that there are some important corner cases where this should not be done. For example when people use OBS to record the desktop. Only one of its streams should be moved to EasyEffects.

There are also other cases were people need to block one particular stream for whatever reason. If WirePlumber starts to move things to EasyEffects by default at some point EasyEffects blocklist will start to not being able to do its job. As well as streams manual handling. Some users prefer that instead.

**Comment by regunakyle** [comment]
Hi, I am using easyeffect for EQ. I also use OBS to stream. 
If I want OBS to record unprocessed audio from my games, is it correct to set `Easy Effect Source` as the audio device in OBS?

**Reply by wwmm** [reply]
> Hi, I am using easyeffect for EQ. I also use OBS to stream. If I want OBS to record unprocessed audio from my games, is it correct to set `Easy Effect Source` as the audio device in OBS?

As far as I remember OBS stream that records your desktop audio through your soundcard monitors is automatically excluded from the output effects pipeline. Your microphone will be processed. This is usually what OBS users want.

If you want your game's output audio to not be processed the game will have to play directly to your soundcard. You can do that by putting the game in EasyEffects blocklist. Or by disabling `Process all output streams` in our preferences window and assuming full manual control of what gets effects or not.

**Reply by regunakyle** [reply]
Thanks for your answer!

**Reply by regunakyle** [reply]
Hi @wwmm , I just did some testing and noticed that OBS (using `Default` audio device) captured the processed audio output from my apps. You can test it by toggling EQ while recording with OBS.
Setting `Easy Effect Sink` as the audio device seems to do what I want (i.e. capturing unprocessed audio).

BTW, are `Easy Effect Sink` and `Easy Effect Source` just the unprocessed input/output?

**Reply by wwmm** [reply]
>BTW, are Easy Effect Sink and Easy Effect Source just the unprocessed input/output?

The output effects pipeline is something like
```
players -> easyeffects sink -> effects -> soundcard
```
. So recording from our virtual sink means you get unprocessed audio. But in  the case of the microphone pipeline we have
```
microphone -> effects -> easyeffects source -> audio recorders
```

**Comment by Nephyrin** [comment]
One thing that is biting me here is that in "process all outputs" mode easyeffects will constantly steal streams that are not going to the device it is managing.  I would expect that if I set easyeffects to use device A, it would be stealing streams for device A and routing them to the easy effects output, but not stealing streams targeting device B.

**Reply by wwmm** [reply]
> I would expect that if I set easyeffects to use device A, it would be stealing streams for device A and routing them to the easy effects output, but not stealing streams targeting device B.

We try to do that. But there are limitations. The only cases where we can reliably do that without side effects are the ones where the audio stream asked the audio server for a specific device through the `PW_KEY_TARGET_OBJECT`tag. The code that checks for this is https://github.com/wwmm/easyeffects/blob/77210c6ec09f75cffb9faca9e3bdf350a9291e88/src/pw_node_manager.cpp#L304C66-L304C86

. If the streams sets that tag and the device in the tag is not the one easyeffects is using we ignore the streams. The problem is that streams are not required to set `PW_KEY_TARGET_OBJECT`. And without this tag there is no good way to know if we should avoid the stream or not.
